pkg/systemd/services/service.jsx

```diff
@@ -40,147 +41,143 @@
 }
 
-export class Service extends React.Component {
-    constructor(props) {
-        super(props);
-
-        this.state = {
-            unit_id: null,
-            unit_props: null,
-            error: null,
-            reloading: false,
-        };
-
-        this.path = null;
-        this.subscriptions = [];
-    }
-
-    async componentDidMount() {
-        const dbus = this.props.dbusClient;
-        try {
-            const [path] = await dbus.call(s_bus.O_MANAGER, s_bus.I_MANAGER, "LoadUnit", [this.props.unitId]);
-            this.path = path;
-        } catch (ex) {
-            this.setState({ error: ex.toString() });
-            return;
-        }
-
-        this.subscriptions.push(dbus.subscribe(
+export const Service = ({ dbusClient, owner, unitId, unitIsValid, addTimerProperties, pinnedUnits }) => {
+    const _path = useRef(null);
+    const [error, setError] = useState(null);
+    const [reloading, setReloading] = useState(false);
+    const [unitProps, setUnitProps] = useState(null);
+
+    const updateProperties = () => {
+        const path = _path.current?.path;
+        if (!path)
+            console.error("Service detail", unitId, "updateProperties called without path");
+        const promises = [dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_UNIT])];
+
+        if (unitId.endsWith(".timer"))
+            promises.push(dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_TIMER]));
+        else if (unitId.endsWith(".socket"))
+            promises.push(dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_SOCKET]));
+
+        Promise.all(promises)
+                .then(replies => {
+                    const unit_props = replies[0][0];
+                    const other_props = replies[1]?.[0];
+
+                    // unwrap variants
+                    for (const key in unit_props)
+                        unit_props[key] = unit_props[key].v;
+
+                    if (unitId.endsWith(".timer"))
+                        addTimerProperties(other_props, unit_props);
+                    else if (unitId.endsWith(".socket"))
+                        unit_props.Listen = other_props.Listen.v;
+
+                    unit_props.path = path;
+
+                    debug("Service detail", unitId, "updated properties:", JSON.stringify(unit_props));
+                    setUnitProps(unit_props);
+                    setError(null);
+                })
+                .catch(ex => setError(ex.toString())); // not-covered: unexpected OS error
+    };
+
+    // load object path and set up PropertiesChanged subscription whenever unitId changes
+    useObject(
+        () => {
+            dbusClient.call(s_bus.O_MANAGER, s_bus.I_MANAGER, "LoadUnit", [unitId])
+                    .then(([path]) => {
+                        debug("Service detail", unitId, "loaded path", path);
+
+                        if (_path.current?.propSub)
+                            _path.current.propSub.remove();
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-094909-6117a6e5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)